# Net foreign factor income Dataset Cleaning

In [1]:
import pandas as pd
import numpy as np

In [2]:
fi = pd.read_csv('../data/uncleaned/foreign_income.csv')
fi.head()

Country Name Country Code  \
0                        Aruba          ABW   
1  Africa Eastern and Southern          AFE   
2                  Afghanistan          AFG   
3   Africa Western and Central          AFW   
4                       Angola          AGO   

                                      Indicator Name  Indicator Code  \
0  Net primary income (Net income from abroad) (c...  NY.GSR.NFCY.CD   
1  Net primary income (Net income from abroad) (c...  NY.GSR.NFCY.CD   
2  Net primary income (Net income from abroad) (c...  NY.GSR.NFCY.CD   
3  Net primary income (Net income from abroad) (c...  NY.GSR.NFCY.CD   
4  Net primary income (Net income from abroad) (c...  NY.GSR.NFCY.CD   

          1960         1961         1962         1963         1964  \
0          NaN          NaN          NaN          NaN          NaN   
1          NaN          NaN          NaN          NaN          NaN   
2  11111037.78  11111128.89  11111128.89  15555515.56  15555515.56   
3          NaN          NaN          NaN          NaN          NaN   
4          NaN          NaN          NaN          NaN          NaN   

         1965  ...          2012          2013          2014          2015  \
0         NaN  ... -1.160894e+08 -1.643575e+08 -1.027374e+08 -1.247486e+08   
1         NaN  ...           NaN           NaN           NaN           NaN   
2  20000000.0  ...  3.181335e+07  7.174855e+07 -2.378156e+06  1.807891e+08   
3         NaN  ...           NaN           NaN           NaN           NaN   
4         NaN  ... -1.017317e+10 -9.660444e+09 -8.335642e+09 -4.434425e+09   

           2016          2017          2018          2019          2020  \
0 -1.352514e+08 -1.705587e+08 -2.147486e+08 -1.532961e+08 -6.810056e+07   
1           NaN           NaN           NaN           NaN           NaN   
2  1.795680e+08  2.486650e+08  1.914091e+08  3.005247e+08  1.889420e+08   
3           NaN           NaN           NaN           NaN           NaN   
4 -2.599234e+09 -4.239218e+09 -6.009502e+09 -6.266159e+09 -4.512744e+09   

           2021  
0 -9.284916e+07  
1           NaN  
2  1.240617e+08  
3           NaN  
4 -5.323940e+09  

[5 rows x 66 columns]

In [3]:
fi.shape

(266, 66)

**This Dataset has about 266 countries records for 62 countries and 3 other features**

In [4]:
fi.nunique()

Country Name      266
Country Code      266
Indicator Name      1
Indicator Code      1
1960               61
                 ... 
2017              199
2018              197
2019              194
2020              187
2021              179
Length: 66, dtype: int64

**The Indicator code and name are constant variables, so the data is only dependent on Country and the year of value.**

**Hence, we drop these features.**

In [5]:
fi = fi.drop(columns = ['Country Name','Indicator Name','Indicator Code'])
fi


Country Code          1960          1961          1962          1963  \
0            ABW           NaN           NaN           NaN           NaN   
1            AFE           NaN           NaN           NaN           NaN   
2            AFG  1.111104e+07  1.111113e+07  1.111113e+07  1.555552e+07   
3            AFW           NaN           NaN           NaN           NaN   
4            AGO           NaN           NaN           NaN           NaN   
..           ...           ...           ...           ...           ...   
261          XKX           NaN           NaN           NaN           NaN   
262          YEM           NaN           NaN           NaN           NaN   
263          ZAF -3.303999e+08 -3.723999e+08 -3.233999e+08 -3.233999e+08   
264          ZMB -7.400000e+07 -6.571429e+07 -7.000000e+07 -6.942857e+07   
265          ZWE -1.909990e+08 -2.351740e+08 -2.685658e+08 -3.127801e+08   

             1964         1965          1966          1967          1968  ...  \
0             NaN          NaN           NaN           NaN           NaN  ...   
1             NaN          NaN           NaN           NaN           NaN  ...   
2    1.555552e+07   20000000.0  2.888889e+07  3.555556e+07  2.888889e+07  ...   
3             NaN          NaN           NaN           NaN           NaN  ...   
4             NaN          NaN           NaN           NaN           NaN  ...   
..            ...          ...           ...           ...           ...  ...   
261           NaN          NaN           NaN           NaN           NaN  ...   
262           NaN          NaN           NaN           NaN           NaN  ...   
263 -3.751999e+08 -415799833.9 -4.339998e+08 -4.493998e+08 -5.165998e+08  ...   
264 -9.871429e+07  -65000000.0 -8.285714e+07 -7.228571e+07 -7.442857e+07  ...   
265 -4.469387e+08 -252962655.7 -1.859027e+08 -1.269006e+08 -1.284504e+08  ...   

             2012          2013          2014          2015          2016  \
0   -1.160894e+08 -1.643575e+08 -1.027374e+08 -1.247486e+08 -1.352514e+08   
1             NaN           NaN           NaN           NaN           NaN   
2    3.181335e+07  7.174855e+07 -2.378156e+06  1.807891e+08  1.795680e+08   
3             NaN           NaN           NaN           NaN           NaN   
4   -1.017317e+10 -9.660444e+09 -8.335642e+09 -4.434425e+09 -2.599234e+09   
..            ...           ...           ...           ...           ...   
261  1.973680e+08  1.614176e+08  1.511544e+08  1.027834e+08  8.247188e+07   
262 -1.569643e+09 -1.815600e+09 -2.621000e+09 -1.345356e+09 -6.412329e+07   
263 -1.088381e+10 -9.691302e+09 -9.453635e+09 -7.943172e+09 -8.291431e+09   
264 -4.319760e+08 -1.148847e+09 -5.519561e+08 -4.120023e+08 -6.471050e+08   
265 -1.868968e+09 -1.870148e+09 -2.195087e+09 -2.358241e+09 -2.546086e+09   

             2017          2018          2019          2020          2021  
0   -1.705587e+08 -2.147486e+08 -1.532961e+08 -6.810056e+07 -9.284916e+07  
1             NaN           NaN           NaN           NaN           NaN  
2    2.486650e+08  1.914091e+08  3.005247e+08  1.889420e+08  1.240617e+08  
3             NaN           NaN           NaN           NaN           NaN  
4   -4.239218e+09 -6.009502e+09 -6.266159e+09 -4.512744e+09 -5.323940e+09  
..            ...           ...           ...           ...           ...  
261  1.438990e+08  1.335368e+08  1.798197e+08  1.868775e+08  1.790331e+08  
262 -4.040135e+07 -1.660698e+07           NaN           NaN           NaN  
263 -1.057583e+10 -1.128881e+10 -9.683840e+09 -5.632082e+09 -8.024928e+09  
264 -1.146814e+09 -4.068053e+08 -4.045870e+08 -5.089758e+08 -1.921844e+09  
265 -1.991136e+09 -7.131709e+08 -3.443985e+08 -4.781034e+08 -6.989881e+08  

[266 rows x 63 columns]

**Now the dataset has countries and years, which we will unpivot to make it fit for feeding it to our model.**

**So, our dataset size before unpivoting and cleaning is 266*62 = 16,492 records**

In [6]:
fi = fi.rename(columns = {"Country Code":"LOCATION"})

In [7]:
fi = fi.melt(id_vars=["LOCATION"], 
        var_name="TIME", 
        value_name="Value")
fi

LOCATION  TIME         Value
0          ABW  1960           NaN
1          AFE  1960           NaN
2          AFG  1960  1.111104e+07
3          AFW  1960           NaN
4          AGO  1960           NaN
...        ...   ...           ...
16487      XKX  2021  1.790331e+08
16488      YEM  2021           NaN
16489      ZAF  2021 -8.024928e+09
16490      ZMB  2021 -1.921844e+09
16491      ZWE  2021 -6.989881e+08

[16492 rows x 3 columns]

In [9]:
fi = fi.dropna()
fi

LOCATION  TIME         Value
2          AFG  1960  1.111104e+07
13         AUS  1960  3.247844e+07
16         BDI  1960 -2.107986e+06
18         BEN  1960 -7.133230e+05
19         BFA  1960  4.451866e+06
...        ...   ...           ...
16486      WSM  2021 -1.538818e+07
16487      XKX  2021  1.790331e+08
16489      ZAF  2021 -8.024928e+09
16490      ZMB  2021 -1.921844e+09
16491      ZWE  2021 -6.989881e+08

[9063 rows x 3 columns]

**After dropping the data boils down to 9063 records**

In [10]:
fi.to_csv('../data/temp/foreign_income_cleaned.csv', mode='wb',index=False)

# EDA

### Statistical Measures of the feature:

In [14]:
fi.describe()

Value
count  9.063000e+03
mean  -4.592678e+08
std    1.482730e+10
min   -1.709660e+11
25%   -8.998640e+08
50%   -8.194407e+07
75%   -3.274721e+05
max    2.929490e+11

### Data Types of the Columns:

In [15]:
fi.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9063 entries, 2 to 16491
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   LOCATION  9063 non-null   object 
 1   TIME      9063 non-null   object 
 2   Value     9063 non-null   float64
dtypes: float64(1), object(2)
memory usage: 283.2+ KB


### Number of rows:

In [17]:
fi.shape[0]

9063